src/main/java/org/elasticsearch/transport/netty/NettyTransport.java

```diff
@@ -1117,6 +1120,45 @@ private void closeChannelsAndWait(Channel[] channels, List<ChannelFuture> future
                     //ignore
                 }
             }
+            for (ChannelFuture future : futures) {
+                future.awaitUninterruptibly();
+            }
+        }
+    }
+
+    class ScheduledPing implements Runnable {
+
+        volatile int successfulPings;
+        volatile int failedPings;
+
+        @Override
+        public void run() {
+            if (lifecycle.stoppedOrClosed()) {
+                return;
+            }
+            for (Map.Entry<DiscoveryNode, NodeChannels> entry : connectedNodes.entrySet()) {
+                // we only support the ping message format since 1.6
+                if (entry.getKey().version().before(Version.V_1_6_0)) {
+                    continue;
+                }
+                for (Channel channel : entry.getValue().allChannels) {
+                    try {
+                        ChannelFuture future = channel.write(NettyHeader.pingHeader());
+                        future.addListener(new ChannelFutureListener() {
+                            @Override
+                            public void operationComplete(ChannelFuture future) throws Exception {
+                                successfulPings++;
```

Comment:
Netty makes sure not to call it concurrently, but agreed, its too hidden, will just change to atomic counters for simplicity